### PR TITLE
Add flag to add_procs to disable threading

### PR DESCRIPTION
**Pull request by katbusch**
For #16729
Adds a flag `disable_threaded_libs` to `addprocs` instead of disabling libs by default and adds a test to exercise the flag

### CODE REVIEW

File: base/docs/helpdb/Base.jl [hunk 1]
```diff
@@ -3016,6 +3016,9 @@ Keyword arguments:
 * `dir`: specifies the working directory on the workers. Defaults to the host's current
          directory (as found by `pwd()`)
 
+* `disable_threaded_libs`: if `true` then certain multithreaded libraries will be set to run on a
+                           single thread. This is useful when using processes instead of threads.
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: could you run `make docs` and commit the modified rst files too?

File: base/multi.jl [hunk 2]
```diff
@@ -1416,7 +1416,9 @@ function addprocs_locked(manager::ClusterManager; kwargs...)
     # some libs by default start as many threads as cores which leads to
     # inefficient use of cores in a multi-process model.
     # Should be a keyword arg?
```
Comment (tkelman) [inline base/multi.jl]: you can delete this comment line now

File: base/multi.jl [hunk 3]
```diff
@@ -1474,7 +1476,8 @@ default_addprocs_params() = AnyDict(
     :topology => :all_to_all,
     :dir      => pwd(),
     :exename  => joinpath(JULIA_HOME,julia_exename()),
-    :exeflags => ``)
+    :exeflags => ``,
+    :disable_threaded_libs => false)
```
Comment (tkelman) [inline base/multi.jl]: I think we might want to leave the default alone, so `true`

File: test/add_procs_threading.jl [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
```
Comment (tkelman) [inline test/add_procs_threading.jl]: note that this file won't get executed unless you add it to the list in `test/choosetests.jl`

File: test/add_procs_threading.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+using Base.Test
+
+type MLKNotImplementedException <: Exception end
+
+"""
+    get_num_threads(n)
+
+Get the number of threads the BLAS library is using.
+"""
+function get_num_threads()
```
Comment (tkelman) [inline test/add_procs_threading.jl]: isn't this available in utils already?

**Comment by tkelman** [comment]
Very nice first PR! Welcome, and thanks!

**Comment by tkelman** [comment]
thinking about this a little more, "disable" arguments are a bit annoying, I think it would be clearer if it were called `enable_threaded_libs`. To match the previous behavior on master it would default to `false`.

**Comment by amitmurthy** [comment]
My thoughts:

- BLAS is the only place it is used currently. Just call it `threaded_blas`? Default is `false`.
- The same should be propagated to the workers too? `init_worker` needs to handle this. Can be passed via a new command-line flag or an environment variable (like `JULIA_WORKER_TIMEOUT`).
- Better to add the tests to `test/parallel_exec.jl` since it adds/removes workers. Need to keep the number of julia processes as low as possible during the main test run to accommodate low-spec'ed CI infrastructure.  Can add to the end of `test/parallel_exec.jl`

**Comment by amitmurthy** [comment]
On second thoughts, for the second point above, we can pass it via the first message sent to the workers, i.e.,  `JoinPGRPMsg` and initialize it in the message handler. And remove it from `init_worker`.

**Comment by amitmurthy** [comment]
Thinking a bit more, this PR could implement finer grained options covering the following scenarios.

1. master(threaded blas), workers(single threaded blas) - scenario when final reduction / processing of collected results are is computed on master but workers are distributed one per core.

2. master(single), workers(single) - current model, master is mostly used as a driver, workers are one per core.

3. master(threaded), workers(threaded) - scenario when workers are launched one per node, the computation can fully leverage threaded blas and this distribution model results in less inter-worker communication. Final reduction/processing on master.

4. master(single), workers(threaded) - similar to 3, but master is mainly a driver.

File: test/add_procs_threading.jl (hunk 5 above)
Comment (kitkat2233) [inline test/add_procs_threading.jl]: Only the set_num_threads versions is wrapped there currently.  I could add this there, though mkl doesn't support it. Would it make sense to do that in a separate PR?

**Comment by kitkat2233** [comment]
Thanks for the feedback. I'll get started on these changes.

Meanwhile taking a step back, I don't know totally why this option in the addprocs function in the first place?  It feels very side-effecty and potentially confusing, leading to #16729.  I wonder if it makes sense to remove it all together and advise users in the docs that they might want to limit BLAS threading.

**Comment by amitmurthy** [comment]
> I wonder if it makes sense to remove it all together and advise users in the docs that they might want to limit BLAS threading.

For users to limit it on the workers would mean an extra explicit `@everywhere` call. Also, initially people would see poorer performance on local `addprocs` before optimizing it.

It does seem useful to change the default to threaded blas on master, with single threads on workers though.

File: base/linalg/blas.jl [hunk 6]
```diff
@@ -112,6 +112,32 @@ function set_num_threads(n::Integer)
     return nothing
 end
 
+type MLKNotImplementedException <: Exception end
```
Comment (tkelman) [inline base/linalg/blas.jl]: Is this supposed to be MKL... ?

**Comment by kitkat2233** [comment]
Thinking about what you said, I think then it makes most sense for addprocs to have a setting for multiprocessing in the child procs it's creating. The master doesn't have the same performance implications.  Right now I've updated it to only give the option for the child processes.  Does that make sense to you too? Otherwise I can add back the master option.

File: base/linalg/blas.jl (hunk 6 above)
Comment (kitkat2233) [inline base/linalg/blas.jl]: Yes!

File: test/parallel_exec.jl [hunk 7]
```diff
@@ -1093,3 +1093,36 @@ let
...
+
+    rmprocs(processes_added)
+
+    BLAS.set_num_threads(4)
+    @test BLAS.get_num_threads() == 4
+    processes_added = addprocs(2, enable_threaded_blas=true)
+    # Master thread default is 4
+    @test BLAS.get_num_threads() == 4
+    # Remote thread default is 2
+    check_remote_num_threads(processes_added, 2)
```
Comment (amitmurthy) [inline test/parallel_exec.jl]: `num_threads` cannot be hardcoded. It is dependent on the number of cores of the machine the test suite is run.

File: base/multi.jl [hunk 8]
```diff
@@ -1479,6 +1481,7 @@ function addprocs_locked(manager::ClusterManager; kwargs...)
 
             if !isempty(launched)
                 wconfig = shift!(launched)
+                wconfig.enable_threaded_blas = Nullable{Bool}(params[:enable_threaded_blas])
```
Comment (amitmurthy) [inline base/multi.jl]: All the other fields of WorkerConfig are initialized in the managers only. This line should be moved here https://github.com/JuliaLang/julia/blob/6b6dead59db79f7940ad2d54d5269e5a319ed796/base/managers.jl#L212 and https://github.com/JuliaLang/julia/blob/6b6dead59db79f7940ad2d54d5269e5a319ed796/base/managers.jl#L323

File: base/managers.jl [hunk 9]
```diff
@@ -79,6 +79,9 @@ Keyword arguments:
 * `dir`: specifies the working directory on the workers. Defaults to the host's current
          directory (as found by `pwd()`)
 
+ * `enable_threaded_blas`: if `true` then  BLAS will run on multiple threads in added
+                           processes. Default is `false`.
+
```
Comment (amitmurthy) [inline base/managers.jl]: Can you add a line here https://github.com/JuliaLang/julia/blob/6b6dead59db79f7940ad2d54d5269e5a319ed796/base/managers.jl#L302 stating that keyword args `dir`, `exename`, `exeflags`, `topology` and `enable_threaded_blas` have the same effect as documented for `addprocs(machines)`

**Comment by ViralBShah** [comment]
If we want this level of control over threading in the 0.5.x release, this PR has to find its way into 0.5.0 since the API changes it introduces are not backwards compatible.

**Comment by tkelman** [comment]
it's too late for that, unfortunately.

File: test/parallel_exec.jl (hunk 7 above)
Comment (kitkat2233) [inline test/parallel_exec.jl]: I'm looking into a more robust way to do this.  It looks like the formula for picking the number of open BLAS threads is pretty complicated, and it's not guaranteed to be using open blas.  Right now I'm doing something that might be flakey: if the number of cores is > 2, assert the number of blas threads is >= 2.  Do you have any suggestions?

**Comment by kitkat2233** [comment]
> If we want this level of control over threading in the 0.5.x release, this PR has to find its way into 0.5.0 since the API changes it introduces are not backwards compatible.
I can keep the old default behavior if backwards compatibility is better--thoughts? I don't have context to figure out how important this is

**Comment by amitmurthy** [comment]
Thanks @kitkat2233. Since it is an API change I guess it is not making it into the 0.5 releases. 

@tkelman is it OK to merge with master now or do we need to wait till 0.5 is released?

File: test/parallel_exec.jl [hunk 10]
```diff
@@ -1093,3 +1093,41 @@ let
...
     @test_throws RemoteException fetch(ref)
 end
+
+# Test addprocs enable_threaded_blas parameter
+function get_remote_num_threads(processes_added)
+    return [remotecall_fetch(BLAS.get_num_threads, proc_id) for proc_id in processes_added]
+end
+
+BLAS.set_num_threads(4)
+@test BLAS.get_num_threads() == 4
```
Comment (tkelman) [inline test/parallel_exec.jl]: wouldn't this fail if you were using the reference blas?

File: base/linalg/blas.jl [hunk 11]
```diff
@@ -112,6 +112,29 @@ function set_num_threads(n::Integer)
     return nothing
 end
 
+"""
+    get_num_threads(n)
+
+Get the number of threads the BLAS library can use.
+"""
+function get_num_threads()
```
Comment (tkelman) [inline base/linalg/blas.jl]: @eschnett you tried to do this in #10028 / #15008 - what was the issue you had in the PR?

It may also be worth making this conditional on openblas version, since I don't think they've had this API for all that long.

File: base/linalg/blas.jl [hunk 12]
```diff
@@ -112,6 +112,29 @@ function set_num_threads(n::Integer)
...
+    if blas == :openblas
+        return ccall((:openblas_get_num_threads, Base.libblas_name), Cint, ())
+    elseif blas == :openblas64
+        return ccall((:openblas_get_num_threads64_, Base.libblas_name), Cint, ())
+    elseif blas == :mkl
+        return ccall((:MKL_Get_Max_Num_Threads, Base.libblas_name), Cint, ())
+    end
+
+    # OSX BLAS looks at an environment variable
+    @static if is_apple()
```
Comment (tkelman) [inline base/linalg/blas.jl]: I don't think the `@static` is necessary

**Comment by tkelman** [comment]
I don't think this is that likely to cause merge conflicts, so it can be merged once it's ready and will work robustly. Since it's not a backport candidate, we could also consider changing the default for 0.6.

File: base/linalg/blas.jl (hunk 11 above)
Comment (kitkat2233) [inline base/linalg/blas.jl]: the vendor() function is in a try, so I copied that technique and return nothing if our known functions fail. is this too much of a hammer?

Comment (tkelman) [inline base/linalg/blas.jl]: Try-catches can be a bit expensive but I'm not sure what better approaches there might be. `Libdl.dlsym` might be a faster path?

Comment (kitkat2233) [inline base/linalg/blas.jl]: Since I've realized the behavior of get_num_threads is more complicated than I thought I'm not going to add it BLAS in this pull request. I stayed with the try in the test.

File: base/managers.jl [hunk 13]
```diff
@@ -107,6 +110,7 @@ This timeout can be controlled via environment variable `JULIA_WORKER_TIMEOUT`.
 The value of JULIA_WORKER_TIMEOUT` on the master process specifies the number of seconds a
 newly launched worker waits for connection establishment.
 """
+
```
Comment (tkelman) [inline base/managers.jl]: don't add the empty line between the docstring and the function

File: test/parallel_exec.jl [hunk 14]
```diff
@@ -1093,3 +1093,84 @@ let
...
+            @test get(worker.config.enable_threaded_blas) == expected
+            return
+        end
+    end
+end
+
+function test_add_procs_threaded_blas()
+    define_get_num_threads()
+    if get_num_threads() == nothing
+        print("Skipping blas num threads tests due to unsupported blas version")
```
Comment (tkelman) [inline test/parallel_exec.jl]: this is worth making a warning

**Comment by kitkat2233** [comment]
Is there anything else I should do here to get this ready for a merge?

**Comment by amitmurthy** [comment]
LGTM.

Will merge in a day if there are no objections.

**Comment by kshyatt** [comment]
@amitmurthy I'm bumping as a merge reminder :)

**Comment by amitmurthy** [comment]
Thanks @kitkat2233 !
